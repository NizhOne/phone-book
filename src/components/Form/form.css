.label-input, .label-input input {
    width: 100%;
}
.add-contact-form {
    display: flex;
    flex-direction: column;
    width: 200px;
    margin: 20px;
}