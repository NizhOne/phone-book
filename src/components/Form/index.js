import { connect } from 'react-redux';

import FormComponent from './Form';
import {addContact, resetId, saveContact} from '../../actions';

const mapDispatchToProps = (dispatch) => {
    return {
        onButtonClick: (contactData) => {
            dispatch(addContact({id: Date.now(), ...contactData}));
        },
        onSaveButtonClick: (id, contactData) => {
            dispatch(saveContact({id,  ...contactData}));
            dispatch(resetId());
        }
    }
};

const mapStateToProps = (state) => {
    let filledData = state.contacts[getDeletingIndexById(state.editData, state.contacts)];
    return {
        isEdit: state.editData,
        formData: state.editData ?
            { name: filledData.name, email: filledData.email, phoneNumber: filledData.phoneNumber } :
            { name: '', email: '', phoneNumber: '' }

    }
};

const getDeletingIndexById = (id, array) => {
    return array.findIndex((item) => {
        return item.id === id
    });
};

const Form = connect(
    mapStateToProps,
    mapDispatchToProps
)(FormComponent);

export default Form;