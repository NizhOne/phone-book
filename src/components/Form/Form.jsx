import React from 'react';
import './form.css';
import PropTypes from 'prop-types';
import { basicForm } from '../../constants';
import Input from '../Input';

class FormComponent extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            formData: props.formData,
        }
    }
    handleInputChange = ({ target: { name, value } }) =>
        this.setState({
            formData: {
                ...this.state.formData,
                [name]: value
            }
        })

    clearInput = () => this.setState({ formData: this.props.formData })

    render() {
        const { isEdit, onButtonClick, onSaveButtonClick } = this.props
        const { formData } = this.state

        return (
            <form className={`add-contact-form`}>
                {basicForm.map(({ type, name, placeholder }) => <Input
                    key={name}
                    type={type}
                    name={name}
                    placeholder={placeholder}
                    value={formData[name]}
                    onChange={this.handleInputChange}
                />
                )}
                <button
                    onClick={(e) => {
                        e.preventDefault();
                        isEdit ?
                            onSaveButtonClick(isEdit, formData) :
                            onButtonClick(formData);
                        this.clearInput()
                    }}>
                    Save
                </button>
            </form>
        );
    }
};

FormComponent.propTypes = {
    formData: PropTypes.object,
    onButtonClick: PropTypes.func,
};

export default FormComponent