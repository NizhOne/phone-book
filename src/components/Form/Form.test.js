import React from 'react';
import Form from './Form';
import renderer from 'react-test-renderer';


test('test', () => {
    const component = renderer.create(
        <Form formData={{name: '123', email: '123', phoneNumber: '123'}} isEdit={true}/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});