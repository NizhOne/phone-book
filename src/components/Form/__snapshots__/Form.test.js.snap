// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test 1`] = `
<form
  className="add-contact-form"
>
  <label
    className="label-input"
  >
    <input
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type={undefined}
      value="123"
    />
  </label>
  <label
    className="label-input"
  >
    <input
      name="phoneNumber"
      onChange={[Function]}
      placeholder="Phone number"
      type={undefined}
      value="123"
    />
  </label>
  <label
    className="label-input"
  >
    <input
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type={undefined}
      value="123"
    />
  </label>
  <button
    onClick={[Function]}
  >
    Save
  </button>
</form>
`;
